### PR TITLE
Limit rlim_max in rlimit_nofile_safe() to nr_open

### LINKED ISSUE #28965: f3f2d02e9731e65eb15c26980305cd47e028f4fc causes generators to fail on v253
### systemd version the issue has been seen with

v253.7-1.4.hs+fb.el9

### Used distribution

CentOS Stream 9

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

Generators run successfully

### Unexpected behaviour you saw

Generators fail to run due to being unable to lower `RLIMIT_NOFILE`. It seems the regression is caused by f3f2d02e9731e65eb15c26980305cd47e028f4fc which was supposed to be just refactoring but by using `FORK_RLIMIT_NOFILE_SAFE`, we now fail to spawn the process if we can't lower `RLIMIT_NOFILE`.

cc @yuwata 

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37124]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37125]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37126]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37127]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37128]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37129]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37130]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37131]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37132]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37133]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37134]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37135]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37136]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37137]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/selinux-autorelabel-generator.sh failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (direxec)[37138]: Failed to lower RLIMIT_NOFILE's soft limit to 1K: Operation not permitted
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-veritysetup-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-bless-boot-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-gpt-auto-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-fstab-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-system-update-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-cryptsetup-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-integritysetup-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-hibernate-resume-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-sysv-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/kdump-dep-generator.sh failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-getty-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-debug-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-rc-local-generator failed with exit status 1.
Aug 24 19:03:12 fna008.01.fnak1.facebook.com (sd-execu[37123]: /usr/lib/systemd/system-generators/systemd-run-generator failed with exit status 1.
```

### PR DESCRIPTION
**Pull request by daandemeyer**
We might inherit a max rlim value that's larger than the kernel's maximum (nr_open). This will cause setrlimit() to fail as the given maximum is larger than the kernel's maximum. To get around this, let's limit the max rlim we pass to rlimit() to the value of nr_open.

Should fix #28965

### CODE REVIEW

**Review by yuwata — APPROVED** [review]
LGTM.

**Comment by yuwata** [comment]
@DaanDeMeyer Feel free to merge this PR if it works.

**Review by keszybz — APPROVED** [review]
LGTM.

**Comment by keszybz** [comment]
Considering that Daan was the OP, let's merge.

File: src/basic/rlimit-util.c [hunk 1]
```diff
@@ -401,7 +401,11 @@ int rlimit_nofile_safe(void) {
         if (rl.rlim_cur <= FD_SETSIZE)
                 return 0;
 
-        rl.rlim_cur = FD_SETSIZE;
+        /* So we might have inherited a hard limit that's larger than the kernel's maximum limit as stored in
+         * /proc/sys/fs/nr_open. If we pass this hard limit unmodified to setrlimit(), we'll get EPERM. To
```
Comment (poettering) [inline src/basic/rlimit-util.c:405]: hmm, how can this actually happen? i mean, whoever set the hard limit we received must have run in the same setrlimit() failure, no? what am i missing?

Comment (daandemeyer) [reply src/basic/rlimit-util.c:405]: How so? I'm pretty sure the kernel hard limit only applies to new processes. So systemd starts and bumps the limit to max, some service lowers the kernel limit, and from then on every process spawned by systemd can't go above that limit. Every new process spawned by systemd does inherit the hard rlimit from systemd which is larger than nr_open.

**Pull request merged**